content/tutorials/reactor-firehose-kinesis.textile

```diff
@@ -0,0 +1,602 @@
+      data: message,
+    })
+
+    const options = {
+      host: "rest.ably.io",
+      port: 443,
+      path: `/channels/${channel}/messages`,
+      method: "POST",
+      headers: {
+        Authorization: `Basic ${Buffer.from(ablyApiKey).toString("base64")}`,
+        "Content-Type": "application/json",
+        "Content-Length": Buffer.byteLength(data),
+      },
+    }
+
+    let req = https.request(options)
+    req.write(data)
+    req.end(null, null, () => {
+      // Request has been sent
+      resolve(req)
+    })
+  })
+}
+
+function add(accumulator, a) {
+  return accumulator + a
+}
+```
+
+This code reads items from the data stream, calculates a running average and then publishes that information as an Ably message to the @processed-sensor-data@ channel.
+
+Publish the function by creating a deployment package for it using the AWS CLI. Execute the following code, replacing the @--role@ argument value with the Role ARN for the @lambda-kinesis-role@ you created earlier in this step:
+
+```[sh]
+cd lambda
+zip function.zip index.js
+aws lambda create-function --function-name ProcessKinesisRecords \
+--zip-file fileb://function.zip --handler index.handler --runtime nodejs12.x \
+--role arn:aws:iam::123456789012:role/lambda-kinesis-role
+```
+
+In the code you wrote above, you are referencing your Ably API key as an environment variable, @ABLY_API_KEY@. In order for this to work, you need to set this environment variable. Execute the following AWS CLI command to achieve this, replacing @<your api key>@ with your Ably API key:
+
+```[sh]
+aws lambda update-function-configuration --function-name ProcessKinesisRecords --environment Variables={ABLY_API_KEY=<your api key>}
+```
+
+Finally, you need to tell your lambda when to execute. In this case, you want it to run when items appear in your data stream, so you need to configure an event source mapping between your lambda and your stream. Run the following command, replacing the argument to @--event-source@ with the ARN of your Kinesis data stream that you created in Step 1.
+
+```[sh]
+aws lambda create-event-source-mapping --function-name ProcessKinesisRecords --event-source arn:aws:kinesis:<region>:<id>:stream/sensor-data-stream --batch-size 5 --starting-position LATEST
+```
+
+With your lambda function ready to calculate the average temperature and humidity readings from your sensors and publish those results to an Ably channel, you now need to subscribe to that channel so that you can display them.
+
+h2. Step 8: Display the sensor data averages
+
+In the final step you will subscribe to the @processed-sensor-data@ channel and display the messages that your lambda publishes to that channel.
+
+First, you need to authenticate the client with Ably. In a production environment you should use "token authentication":/core-features/authentication#token-authentication to achieve this. For this sample application, you will just receive the Ably API key from the server.
```

Comment:
Given the server has an Ably client, this may as well do the appropriate thing and return a token, for example:

```js
app.get("/auth", async (request, response) => {
  ably.auth.requestToken((err, tokenDetails) => {
    if(err) {
      console.error(err)
      response.sendStatus(500)
      return
    }
    response.json(tokenDetails)
  })
})
```

This has the added benefit of being compatible with the `authUrl` client option, so the client can be initialised as just `new Ably.Realtime({ authUrl: '/auth' })` and the SDK will take care of fetching the token from this URL.

Reply:
Yes, not sure what I was thinking here 😊  Done!